model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java

```diff
@@ -391,38 +384,31 @@ public void setEmailVerified(boolean verified) {
 
         List<Predicate> predicates = new ArrayList<>();
         predicates.add(builder.equal(root.get("user"), getEntity()));
-        if (Objects.nonNull(search) && !search.isEmpty()) {
-            Join<UserGroupMembershipEntity, GroupEntity> join = root.join("group");
-            predicates.add(builder.like(join.get("name"), builder.literal("%" + search + "%")));
-        }
 
         queryBuilder.select(builder.count(root));
         queryBuilder.where(predicates.toArray(new Predicate[0]));
         return em.createQuery(queryBuilder);
     }
 
-    private Stream<GroupModel> getGroupModels(Stream<String> groupIds) {
-        return groupIds.map(realm::getGroupById);
-    }
-
     @Override
     public Stream<GroupModel> getGroupsStream() {
-        return closing(getGroupModels(createGetGroupsQuery(null, null, null).getResultStream()));
+        return getGroupsStream(null, null, null);
     }
 
     @Override
-    public Stream<GroupModel> getGroupsStream(String search, int first, int max) {
-        return closing(getGroupModels(createGetGroupsQuery(search, first, max).getResultStream()));
+    public Stream<GroupModel> getGroupsStream(String search, Integer first, Integer max) {
+        return session.groups().getGroupsStream(realm, closing(createGetGroupsQuery().getResultStream()), search, first, max);
     }
 
     @Override
     public long getGroupsCount() {
-        return createCountGroupsQuery(null).getSingleResult();
+        return createCountGroupsQuery().getSingleResult();
     }
 
     @Override
     public long getGroupsCountByNameContaining(String search) {
-        return createCountGroupsQuery(search).getSingleResult();
+        if (search == null) return getGroupsCount();
+        return getGroupsStream(search, null, null).count();
```

Comment:
@hmlnarik @vramik Should we have a count query here? This would mean we need another count method in GroupProvider which is matching this method: `session.groups().getGroupsStream(realm, closing(createGetGroupsQuery().getResultStream()), search, first, max)`

Reply:
Yes